Userland/Libraries/LibSQL/BTree.h

```diff
@@ -0,0 +1,293 @@
+    [[nodiscard]] u32 size() const { return m_entries.size(); }
+    [[nodiscard]] Vector<TreeNodeEntry<KeyType>> nodes() const { return m_entries; }
+
+    void insert(KeyType const &key, u32 data_pointer)
+    {
+        TreeNodeEntry<KeyType> entry(0, key, data_pointer);
+        if (size() == max_entries()) {
+            split_and_insert(entry);
+        } else {
+            just_insert(entry);
+        }
+    }
+
+    [[ nodiscard ]] bool is_leaf() const { return m_down_pointers.is_empty(); }
+
+    [[ nodiscard ]] int max_entries() const { return 4; } // FIXME
+
+private:
+    bool m_is_leaf { true };
+    RefPtr<BTree<KeyType>> m_tree;
+    TreeNode<KeyType> *m_up;
+    u32 m_pointer;
+    Vector<TreeNodeEntry<KeyType>> m_entries;
+    Vector<u32> m_down_pointers;
+    Vector<TreeNode<KeyType> *> m_down_nodes;
+
+    void insert_entry(TreeNodeEntry<KeyType> const &node, TreeNode<KeyType> *left, TreeNode<KeyType> *right)
+    {
+        VERIFY(m_entries.size() < max_entries());
+    }
+
+    void just_insert(TreeNodeEntry<KeyType> &entry) {
+        VERIFY(size() < max_entries());
+        for (auto ix = 0; ix < size(); ix++) {
+            if (entry < m_entries[ix]) {
+                m_entries.insert(ix, entry);
+                return;
+            }
+        }
+        m_entries.append(entry);
+    }
+
+    void split_and_insert(TreeNodeEntry<KeyType> &entry)
+    {
+        if (!m_up) {
+            // Make new m_up. This is the new root node.
+        }
+
+        // Find median:
+        TreeNodeEntry<KeyType> median;
+        auto *new_node = new TreeNode<KeyType>(m_tree, m_up, 0);
+        auto move_right = max_entries()/2;;
+        if (m_entries[max_entries()/2] < entry) {  // New entry belongs in new, right, node.
+            median = m_entries[max_entries()/2];
+            move_right--;
+            new_node->insert_entry(entry);
+        } else if (m_entries[max_entries()/2 - 1] > entry) { // New entry belongs in current node,
+                                                             // which will be the new left node.
+            median = m_entries.take(max_entries()/2 - 1);
+            just_insert(entry);
+        } else {                                             // New entry is the median.
+            median = entry;
+        }
+        for (auto i = 0; i < move_right; i++) {
+            new_node->insert_entry(m_entries.take_last());
+        }
+        m_up->insert_entry(median, /* left */ this, /* right */ new_node);
+    }
+
+    friend int deserialize(TreeNode<KeyType> &, ByteBuffer const &, int);
+};
+
+template <typename KeyType>
+ByteBuffer serialize(TreeNode<KeyType> const &leaf)
+{
+    ByteBuffer buffer(serialize(leaf.size()));
+    for (auto &node : leaf.nodes()) {
+        buffer += serialize(node);
+    }
+    return buffer;
+}
+
+template <typename KeyType>
+int deserialize(TreeNode<KeyType> &leaf, ByteBuffer const &buffer, int offset)
```

Comment:
Thanks- will look into.

Reply:
reading again I now see you probably meant 'tree' instead of 'free', and in that case let me explain. The tree lives in persistent storage. As you traverse it to find keys or iterate over all keys, you need to load nodes from storage, and if you insert keys, update what they point to, or delete keys, you need flush the new tree structure to storage.

This is done using the serialization and deserialization methods.

Reply:
I actually meant "free" - as in not part of a specific class.

Reply:
Ah OK. They're mostly gone; I only have template overrides for basic types now.